### PR TITLE
Fix a bug in function removeEmptyNodes and update testcases

### PR DESCRIPTION
**Pull request by fanwgwg**
Function `removeEmptyNodes` will take in an array of `ItemData`, and remove all nodes that have no child items (including grandchild items).

### Reviewer
@Randy-Ma

### CODE REVIEW

File: src/LibraryUtilities.ts [hunk 1]
```diff
@@ -510,13 +510,21 @@ export function buildLibrarySectionsFromLayoutSpecs(loadedTypes: any, layoutSpec
 
 // Remove empty non-leaf nodes from items
 export function removeEmptyNodes(items: ItemData[]) {
+    let itemRemoved = false;
+
     items.forEach((item, index, items) => {
         if (item.childItems.length > 0) {
-            removeEmptyNodes(item.childItems);
+            if (removeEmptyNodes(item.childItems)) {
+                removeEmptyNodes(items);
```
Comment (Randy-Ma) [inline src/LibraryUtilities.ts]: Unnecessary computation is happening here. Assuming there are 1000 items, the last one has some child items removed, then all the first 999 items will be tried again.

File: src/LibraryUtilities.ts [hunk 2]
```diff
@@ -510,13 +510,21 @@ export function buildLibrarySectionsFromLayoutSpecs(loadedTypes: any, layoutSpec
 
 // Remove empty non-leaf nodes from items
 export function removeEmptyNodes(items: ItemData[]) {
+    let itemRemoved = false;
+
     items.forEach((item, index, items) => {
         if (item.childItems.length > 0) {
-            removeEmptyNodes(item.childItems);
+            if (removeEmptyNodes(item.childItems)) {
+                removeEmptyNodes([item]);
```
Comment (Randy-Ma) [inline src/LibraryUtilities.ts]: I am pretty sure that this won't actually remove item from the original items array.

Comment (fanwgwg) [inline src/LibraryUtilities.ts]: This is actually still passing by reference, and I've tried and tested that it works.

File: src/LibraryUtilities.ts [hunk 3]
```diff
@@ -510,13 +510,21 @@ export function buildLibrarySectionsFromLayoutSpecs(loadedTypes: any, layoutSpec
...
+            }
         } else if (item.itemType === "section" || item.itemType === "category" || item.itemType === "group") {
             items.splice(index, 1);
+            itemRemoved = true;
+            removeEmptyNodes(items);
```
Comment (Randy-Ma) [inline src/LibraryUtilities.ts]: If we remove this line, it does not work?

File: src/LibraryUtilities.ts (hunk 2 above)
Comment (Randy-Ma) [inline src/LibraryUtilities.ts]: Hmmmm

Comment (Randy-Ma) [inline src/LibraryUtilities.ts]: I mean item will be removed from the new array - [item], but not from the original array.

**Pull request merged**